File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.cpp
class Solution {
public:
    int total_bouquets(vector<int>& days,int wait_days,int k){
        int total=0;
        for(int i=0;i<days.size();i++){
            if(days[i]>wait_days) continue;
            else{
                int anc=i;
                bool flag=false;
                for(;i<days.size();i++){
                    if(days[i]>wait_days) break;
                    if(i-anc+1==k){
                        total++;
                        break;
                    }
                }
                // for(;i<days.size() && (i-anc+1)<k;i++){
                //     if(days[i]>wait_days){
                //         flag=true;
                //         cout<<i<<endl;
                //         break;
                //     }
                // }
                // if(!flag && days[i]<=wait_days && (i-anc+1)==k){
                //     cout<<anc<<"-"<<i<<endl;
                //     total++;
                // } 
            }
        }
        // cout<<wait_days<<" "<<total<<endl;
        return total;
    }
    int minDays(vector<int>& bloomDay, int m, int k) {
        // if(k>bloomDay.size()) return -1;
        int l=1,r=*max_element(begin(bloomDay),end(bloomDay)),ans=-1; 
        while(l<=r){
            int mid = l+(r-l)/3;
            if(total_bouquets(bloomDay,mid,k)>=m){
                ans=mid;
                r=mid-1;
            }
            else l=mid+1;
        }
        return ans;
    }
};

File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.py
class Solution:
    def getBoquet(self,bloomDay,days,k): 
        anc,cnt=-1,0 
        i=0
        while i<len(bloomDay): 
            if bloomDay[i]<=days:
                anc=i
                while i<len(bloomDay) and bloomDay[i]<=days: 
                    if i-anc+1==k: 
                        cnt+=1
                        break 
                    i+=1
            i+=1
        return cnt
        

    def minDays(self, bloomDay: List[int], m: int, k: int) -> int:

        # 1.  date range - 1 - max_val(bloomDay)
        # 2. for each day you figure out if the the flowers bloomed can form boquet 
        # 3. return the first the day which it blooms 
        l,r=1,max(bloomDay)
        print(r)
        res = -1 
        while l<=r: 
            mid = l+(r-l)//2 
            print(mid)
            if self.getBoquet(bloomDay,mid,k)>=m:
                res=mid
                r=mid-1 
            else: 
                l=mid+1 
        return res 